cgp/genome.py

```diff
@@ -134,6 +137,78 @@ def __repr__(self) -> str:
         s += ")"
         return s
 
+    def determine_permissible_value_per_gene(self, gene_idx: int, region_idx: int) -> np.array:
+
+        if self._is_input_region(region_idx):
+            return self._determine_permissible_values_input_gene(gene_idx, region_idx)
+
+        elif self._is_hidden_region(region_idx):
+            return self._determine_permissible_values_hidden_gene(gene_idx, region_idx)
+
+        elif self._is_output_region(region_idx):
+            return self._determine_permissible_values_output_gene(gene_idx)
+
+        else:
+            assert False  # should never be reached
+
+    def determine_permissible_values(self) -> List[np.array]:
+        """Determine permissible values for every gene.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        ----------
+        permissible_values
+            List[numpy array]: List of numpy arrays over permissible values for every gene
+        """
+        permissible_values = []
+        for gene_idx in range(self._n_genes):
+            region_idx = gene_idx // self._length_per_region
+            permissible_values_per_gene = self.determine_permissible_value_per_gene(
+                gene_idx, region_idx
+            )
+            permissible_values.append(permissible_values_per_gene)
+        return permissible_values
+
+    def _determine_permissible_values_input_gene(self, gene_idx: int, region_idx: int) -> np.array:
+
+        if region_idx == 0:
+            region_idx += self._length_per_region  # avoid issues with x % 0
```

Comment:
hmm, what does this block do?

Reply:
Below (line 180) i check whether it is the first gene in the region: 
```python
 if gene_idx % (region_idx * self._length_per_region) == 0:

```
and the % operator can not be used with `region_idx = 0`. Therefore I pretend I am at unit 1 instead of at unit 0. (it is a bit a dirty hack i know, but i did for the lack of a better idea)

Reply:
ah, i see. i would remove this and adjust the if statement below like such:
```python
 if gene_idx % (max(1, region_idx) * self._length_per_region) == 0:
```
but that's just a personal preference. either way it needs to be clearly documented :)

Reply:
ah, there's actually a function for this: `self._is_function_gene(gene_idx)` should do the trick. please also fix this in the other `if`s